deepchem/models/torch_models/ferminet.py

```diff
@@ -497,15 +529,64 @@ def train(self,
         weight_decay: float (default: 0.0001)
             contains the weight_decay for the model fitting
         """
+
+        # hook function below is an efficient way modifying the gradients on the go rather than looping
+        def energy_hook(grad, random_walk_steps):
+            """
+            hook function to modify the gradients
+            """
+            # using non-local variables as a means of parameter passing
+            nonlocal energy_local, energy_mean
+            new_grad = (2 / random_walk_steps) * (
+                (energy_local - energy_mean) * grad).float()
+            return new_grad
+
         optimizer = torch.optim.Adam(self.model.parameters(),
                                      lr=lr,
                                      weight_decay=weight_decay)
+
         if (self.tasks == 'pretraining'):
             for _ in range(nb_epoch):
                 optimizer.zero_grad()
+                self.loss_value = torch.tensor(0.0)
                 self.molecule.move()
-                self.loss_value = (torch.mean(self.model.running_diff) /
-                                   self.random_walk_steps)
                 self.loss_value.backward()
                 optimizer.step()
                 self.model.running_diff = torch.zeros(self.batch_no)
+
+        if (self.tasks == 'training'):
```

Comment:
if "training", the gradients are modified based on this equation
![grad](https://github.com/deepchem/deepchem/assets/80086120/fee06cf9-3b0d-4d6a-b8fc-a8089e4cf9fd)
will add this as a comment in latex(writing for blog as well)